docs/development/architecture.org

```diff
@@ -0,0 +1,130 @@
+
+The frontend (=nixui/graphics=) constructs various widgets comprising Nix-Gui and loads them with data from the backend. The frontend doesn't manage any edit state. All "edit" actions performed in the GUI send a request-for-update to the backend.
+
+* Backend
+The backend handles three major tasks: configuration retrieval to initialize data-structures, in-memory configuration mutation, and configuration commitment. It provides APIs for frontend interaction, but doesn't concern itself with how the frontend renders any of its data.
+
+** Loading the Option Tree Data-Structure
+Nix-Gui queries information from the configuration path and =<nixpkgs>= in order to construct a data-structure (=nixui.options.option_tree.OptionTree=) mapping options and attributes to metadata (=nixui.options.option_tree.OptionData=) including the attributes =NixType= and =Optiondefinition=:
+- =_type=: A static python-object representation of the option type (=NixType=). Informs the frontend of how =FieldWidget= should be generated and input checking should be performed.
+- =system_default_definition=: Provided by =<nixpkgs>=, the default value if nothing is set in the configuration path.
+- =configured_definition=: Provided by the configuration path, overrides =system_default_definition=.
+- =in_memory_definition=: Undefined when the data-structure is loaded, can be mutated by the user and later committed to disk.
+
+=nixui/nix/lib.nix= (invoked by =nixui.options.nix_eval=) contains the function =get_all_nixos_options()= which is used to retrieve =OptionData= including =system_default_definition= and =_type= for the universe of NixOS options.
+
+=nixui/options/parser.py= contains the function =get_all_option_values(module_path)= which
+- retrieves all =attributePath = string representation of nix expression;= pairs in the file by parsing the =nixui.options.syntax_tree.SyntaxTree= constructed using =nix_dump_syntax_tree_json=.
+- resolves the path of all =imports= and recurses
+
+The result of =get_all_option_values(module_path)= is used to set =OptionData.configured_definition= where applicable.
+
+*** Caveat
+The =OptionTree= has a node for each option in the option tree, as well as attributes that aren't actually options.
+
+For example =users.ldap.loginPam= is a valid nix option. However its "parent", =users.ldap= isn't a valid option. Regardless, =users.ldap= has its own node in the =OptionTree=.
+
+Likewise, =environment.etc."resolv.conf".text= isn't an option, it is an attribute of the option =environment.etc=. It will have a node in the =OptionTree=. This is true of =attribute set of submodules=, as well as =submodule= type options.
+
+=list of= options are part of the option tree despite not having a proper attribute path, e.g. a list of allowed TCP ports in the firewall will have the element
+#+begin_src nix
+networking.firewall.allowedTCPPorts."[0]"
```

Comment:
we risk collisions in rare cases ... what about `allowedTCPPorts."__nixgui_idx 0"`

Reply:
I was thinking of having `networking.firewall.allowedTCPPorts[0]` since this isn't a legal option path in Nix. Thoughts?

Reply:
Actually `networking.firewall."allowedTCPPorts[0]"` is a valid option path I think, however the risk of collision seems low.

Reply:
yepp, arrays should work

Reply:
https://github.com/nix-gui/nix-gui/issues/218

Reply:
wait, what are we talking about here?
how is the dataflow between configuration.nix and python datastructures?
sounds like there is double parsing via json ... 
ideally we should use python bindings for rnix-parser

https://duckduckgo.com/?q=swig+rust+python&ia=web

Reply:
but then, this is probably not the bottleneck of nix-gui, so zero priority